compiler/hash-ir/src/ir.rs

```diff
@@ -117,77 +129,183 @@ pub enum Mutability {
+}
+
+/// A defined statement within the IR
+#[derive(Debug, PartialEq, Eq)]
+pub enum StatementKind<'i> {
     /// Filler kind when expressions are optimised out or removed for other
     /// reasons.
     Nop,
     /// A constant value.
     Const(Const),
+
+    /// A local variable value
+    Local(Local),
+
     /// An identity expression, essentially wrapped with an inner expression.
-    Identity(&'i Expr<'i>),
+    Identity(&'i Statement<'i>),
     /// A unary expression with a unary operator.
-    Unary(UnaryOp, &'i Expr<'i>),
+    Unary(UnaryOp, &'i Statement<'i>),
     /// A binary expression with a binary operator and two inner expressions.
-    Binary(BinOp, &'i Expr<'i>, &'i Expr<'i>),
-    /// A function call with a number of specified arguments.
-    Call(&'i Expr<'i>, &'i [Expr<'i>]),
+    Binary(BinOp, &'i Statement<'i>, &'i Statement<'i>),
     /// An index expression e.g. `x[3]`
-    Index(&'i Expr<'i>, &'i Expr<'i>),
+    Index(&'i Statement<'i>, &'i Statement<'i>),
     /// An assignment expression, a right hand-side expression is assigned to a
     /// left hand-side pattern e.g. `x = 2`
-    Assign(&'i Pat<'i>, &'i Expr<'i>),
-    /// An expression which represents a re-assignment to a pattern with a right
-    /// hand-side expression and a binary operator that combines assignment
-    /// and an operator, e.g. `x += 2`
-    AssignOp(&'i Pat<'i>, BinOp, &'i Expr<'i>),
+    Assign(Local, &'i Statement<'i>),
+
     /// An expression which is taking the address of another expression with an
     /// mutability modifier e.g. `&mut x`.
-    AddrOf(Mutability, &'i Expr<'i>),
+    Ref(Mutability, &'i Statement<'i>, AddressMode),
+
+    /// Allocate some value on the the heap using reference
+    /// counting.
+    Alloc(Local),
 }
 
-/// Essentially a block
+/// [Terminator] statements are essentially those that affect control
+/// flow.
 #[derive(Debug, PartialEq, Eq)]
-pub struct Body<'i> {
-    exprs: &'i [Expr<'i>],
+pub struct Terminator {
+    /// The kind of [Terminator] that it is.
+    pub kind: TerminatorKind,
+    /// The [Span] of the statement, relative to the [Body]
+    /// `source-id`.
+    pub span: Span,
 }
 
 #[derive(Debug, PartialEq, Eq)]
-pub enum Ty {
-    USize,
-    Bool,
-    U8,
-    U16,
-    U32,
-    U64,
-    ISize,
-    I8,
-    I16,
-    I32,
-    I64,
-    F32,
-    F64,
-    Char,
-    Void,
+pub enum TerminatorKind {
+    /// A simple go to block directive.
+    Goto(BasicBlock),
+
+    /// Return from the parent function
+    Return,
+
+    /// Perform a function call
+    Call {
+        // op: todo!(),
+        args: Vec<Local>,
+
+        /// Where to return after completing the call
+        target: Option<BasicBlock>,
+    },
+
+    /// Denotes that this terminator should never be reached, doing so will
+    /// break IR control flow invariants.
+    Unreachable,
+
+    /// Essentially a `jump if <0> to <1> else go to <2>`
+    Switch(Local, Vec<(Const, BasicBlock)>, BasicBlock),
```

Comment:
Could this be smallvec? Or maybe &'i slice?

Reply:
Could be indeed